Python/pytime.c

```diff
@@ -781,35 +869,71 @@ _PyTime_GetMonotonicClockWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
 
 
 #ifdef MS_WINDOWS
+static int
+init_win_perf_counter(LONGLONG *t0, LONGLONG *perf_frequency,
+                      _PyTime_t *num_p, _PyTime_t *den_p)
+{
+    LARGE_INTEGER now;
+    LARGE_INTEGER freq;
+    _PyTime_t num, den, gcd;
+
+    QueryPerformanceCounter(&now);
+    if (!QueryPerformanceFrequency(&freq)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    if (freq.QuadPart < 1) {
+        return -1;
+    }
+
+    *t0 = now.QuadPart;
+    *perf_frequency = freq.QuadPart;
+
+    /* Compute the greatest common divisor to minimize numerator and
+       denominator, to reduce the risk of integer overflow in
+       _PyTime_GetWinPerfCounterWithInfo(). */
+    num = SEC_TO_NS;
+    den = freq.QuadPart;
+    gcd = _PyTime_GCD(num, den);
+    num /= gcd;
+    den /= gcd;
+    if (num < 1 || den < 1) {
+        return -1;
+    }
+
+    *num_p = num;
+    *den_p = den;
+    return 0;
+}
+
 int
-_PyTime_GetWinPerfCounterWithInfo(_PyTime_t *t, _Py_clock_info_t *info)
+_PyTime_GetWinPerfCounterWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
 {
-    static LONGLONG cpu_frequency = 0;
-    static LONGLONG ctrStart;
+    static LONGLONG perf_frequency = 0;
+    static _PyTime_t numer = 0, denom = 0;
+    static LONGLONG t0 = 0;
     LARGE_INTEGER now;
-    double diff;
+    _PyTime_t t;
 
-    if (cpu_frequency == 0) {
-        LARGE_INTEGER freq;
-        QueryPerformanceCounter(&now);
-        ctrStart = now.QuadPart;
-        if (!QueryPerformanceFrequency(&freq) || freq.QuadPart == 0) {
-            PyErr_SetFromWindowsErr(0);
+    if (perf_frequency == 0) {
+        if (init_win_perf_counter(&t0, &perf_frequency, &numer, &denom) < 0) {
```

Comment:
Yeah, it's on purpose.

_PyTime_Init() makes sure the time.time(), time.monotonic() and time.perf_counter() clocks work. If a function returns -1 without raising an exception, it's ok. _Py_InitializeMainInterpreter() calls Py_FatalError() with a nice error message if _PyTime_Init() fails. If an exception is raised, it is displayed. If no exception is raised, we only have the Py_FatalError() message which should be enough.

I only uesd the -1 without any exception set for cases which "should never occur in practice" :-)

Reply:
Why not use asserts instead?

It would be better to document this behavior explicitly, because it is different from conventional behavior of C API.

Reply:
> Why not use asserts instead?

I prefer to not make too many assumptions on clock. Some platforms are really weird.

When I implemented time.monotonic(), I added a check at runtime, but only in debug mode, to really make sure that the clock is monotonic. The test failed on a strange Debian VM. We decided to just drop the assertion... Honestly, I only added the assertion to really check if some weird clock exist in the world. So yes, they do exist :-)

Reply:
> It would be better to document this behavior explicitly, because it is different from conventional behavior of C API.

In fact, the same C function is used in 3 different ways:

* _PyTime_Init(): read each clock once to check if the clock works. The check is required since clock_gettime(CLOCK_MONOTONIC) can return an error on Hurd for example.
* time.get_clock_time()
* time.time(), time.perf_counter(), time.monotonic()